### ISSUE TITLE
Correspondence between registry and .py file

### ISSUE THREAD

**Issue by Lg955** [csuhan/s2anet#90]
There are 7 Registries in the old S2anet model:
`BACKBONES = Registry('backbone')`
`NECKS = Registry('neck')`
`ROI_EXTRACTORS = Registry('roi_extractor')`
`SHARED_HEADS = Registry('shared_head')`
`HEADS = Registry('head')`
`LOSSES = Registry('loss')`
`DETECTORS = Registry('detector')`.

After my test, the BACKBONES corresponds to `/mmdet/models/backbones/resnet.py` ； the  NECKS  corresponds to `/mmdet/models/necks/FPN.py`;   the DETECTORS corresponds to `/mmdet/models/detectors/s2anet.py`. 

But I can't find the  ROI_EXTRACTORS、 SHARED_HEADS 、HEADS  correspond to which .py file, so I want to get some help.

**Comment by csuhan** [comment]
https://github.com/csuhan/s2anet/tree/master/mmdet/models/roi_extractors
https://github.com/csuhan/s2anet/tree/master/mmdet/models/shared_heads
https://github.com/csuhan/s2anet/tree/master/mmdet/models/anchor_heads

**Comment by Lg955** [comment]
Thank you again! 
In the `mmdet/models/anchor_heads/s2anet_head.py` , I think the code of calculating offset does not match formula 3 and 4.
https://github.com/csuhan/s2anet/blob/dce96b7aeee170a44a7a2518ec9d684377da287a/mmdet/models/anchor_heads/s2anet_head.py#L59
As follows:


        def _calc_offset(anchors, kernel_size, featmap_size, stride):
            dtype, device = anchors.dtype, anchors.device
            feat_h, feat_w = featmap_size
            pad = (kernel_size - 1) // 2
            idx = torch.arange(-pad, pad + 1, dtype=dtype, device=device)   #idx=[-1,0,1]
            yy, xx = torch.meshgrid(idx, idx)                               #1 xx-yy is the r in Formula 3
            xx = xx.reshape(-1)
            yy = yy.reshape(-1)

            # get sampling locations of default conv
            xc = torch.arange(0, feat_w, device=device, dtype=dtype)        #2 xc-yc is the each point in feature map
            yc = torch.arange(0, feat_h, device=device, dtype=dtype)
           
            yc, xc = torch.meshgrid(yc, xc)
            xc = xc.reshape(-1) 
            yc = yc.reshape(-1)  
            x_conv = xc[:, None] + xx                                       #3  x_conv-y_conv is the (p+r) in Formula 4
            y_conv = yc[:, None] + yy  

            # get sampling locations of anchors
            
            x_ctr, y_ctr, w, h, a = torch.unbind(anchors, dim=1)            #4  (x_ctr, y_ctr, w, h, a) represent the corresponding anchor box at location p 
            x_ctr = x_ctr / stride                                          #5  x_ctr is the Center point coordinates 
            y_ctr = y_ctr / stride  
            w_s = w / stride        
            h_s = h / stride       
            
            cos, sin = torch.cos(a), torch.sin(a)  
            dw, dh = w_s / kernel_size, h_s / kernel_size                   #6 it represents the 1/k*(w, h) in Formula 3
            x, y = dw[:, None]*xx, dh[:, None]*yy                           #7 1/k*(w, h)*r  
    
            xr = cos[:, None]*x-sin[:, None]*y                              #8 (1/k*(w,h)*r)*R(theta)  
            yr = sin[:, None]*x+cos[:, None]*y                              #9 (1/k*(w,h)*r)*R(theta)
    
            x_anchor, y_anchor = xr+x_ctr[:, None], yr+y_ctr[:, None]       #10 represents the L_p^r in Formula 4
                                                                             but the x_ctr-y_ctr is not multiplied by the angle R(theta)
           
            # get offset filed
            offset_x = x_anchor - x_conv                                    #11 represents the Offset(O) in Formula 4
            offset_y = y_anchor - y_conv    
        
            
            # x, y in anchors is opposite in image coordinates,
            # so we stack them with y, x other than x, y
            offset = torch.stack([offset_y, offset_x], dim=-1)
            # NA,ks*ks*2
            offset = offset.reshape(anchors.size(
                0), -1).permute(1, 0).reshape(-1, feat_h, feat_w)
            return offset
  

![340](https://user-images.githubusercontent.com/69025725/125181002-3b3d1380-e233-11eb-8966-11a3bff9df7d.png)
Just as the code comment"# 1 - # 11".

In "# 8 and 9", the xr-yr has undergone an angular transformation (`xr = cos[:, None]*x-sin[:, None]*y `),  but x_ctr-y_ctr hasn't go through the same step in "# 5 - # 10"(the x_ctr-y_ctr just divided by stride). 

So it is different from the Formula 3 in the paper.

**Comment by csuhan** [comment]
Yes, you are right. There is a mistake.

What we want to do in AlignConv is adjusting the conv's sampling locations based on the corresponding anchor location. 
In Eq.3, **x** does not need the angle translation, while `(w, h)` does.
So the correct formula should be:

![image](https://user-images.githubusercontent.com/17927516/125225921-a86baa00-e302-11eb-87b6-057686f57dac.png)


Our code's implementation is correct.

For our paper, this happened when we translate our `long` version formula to the `short` one (current version with verctor representation).

Thank you for your suggestion. We will revise our paper as soon.
